ceno_zkvm/src/tables/ram/ram_impl.rs

```diff
@@ -367,8 +374,8 @@ impl<DVRAM: DynVolatileRamTable + Send + Sync + Clone> DynVolatileRamTableConfig
         num_witness: usize,
         final_mem: &[MemFinalRecord],
     ) -> Result<RowMajorMatrix<F>, ZKVMError> {
-        assert!(final_mem.len() <= DVRAM::max_len(&self.params));
-        assert!(DVRAM::max_len(&self.params).is_power_of_two());
+        assert!(final_mem.len() <= DVRAM::max_len(&self.params, self.addr.multi_factor));
+        assert!(DVRAM::max_len(&self.params, self.addr.multi_factor).is_power_of_two());
         let mut final_table = RowMajorMatrix::<F>::new(final_mem.len(), num_witness);
```

Comment:
We define `assign_structural_instances` to make thing isolated which is good.
But it imply we need to go through witness twice and bring overhead of efficiency.

I suggest a design by defining a new function argument num_structural_witness: usize from here
https://github.com/scroll-tech/ceno/blob/master/ceno_zkvm/src/structs.rs#L316
and we make

```rust
let mut final_table = RowMajorMatrix::<F>::new(final_mem.len(), num_witness + num_structural_witness); 
```

here, so they can assign in same place.

Then we extract & separate witness from here https://github.com/scroll-tech/ceno/blob/master/ceno_zkvm/src/scheme/prover.rs#L99-L121

Reply:
However, the id of the WitIn and StructuralWitIn are currently calculated separately. If they are placed in the same matrix, the `id` will be put together, resulting in needing to mark which `id` represents the `witness` and which ones are the `structure witness`. Wouldn't that be more complicated?

Reply:
We can make this logic just on assignment. It will not affect constrain system and both id still maintain separately.

```rust
pub fn assign_instances<F: SmallField>(...) {
...
// concat structural witness behind witness for efficient representation
// this assure both are same type so we dont need special mark
let addr = StructuralWitIn { 
    id: self.addr + num_witness,
    ... self.addr 
};
...

final_table
            .par_iter_mut()
            ... {
                // take `addr` here instead of `self.addr`
            }    
}
```

Matrix is just an intermediate representation, so we can tolerate some compact format with simpler and unified api and also keep same efficiency.